### PR TITLE
Going deeper on fork orchestration

### PR DESCRIPTION
**Pull request by tiagofilipe12**
This PR intends to at the very least increase the level of forkception, i.e., the ability to had forks inside forks.
For now it just handles one fork inside another fork, but has no capacity to handle a third fork. Also and importantly, fork can proceed a fork, i.e. we can have something like this:

```javascript
const pipeline3 = join(
  task0,
  fork(
      fork(
        task1,
        task3
      ),
      task6
  ),
  task5
)
```

### CODE REVIEW

File: lib/orchestrators/join.js [hunk 1]
```diff
@@ -32,19 +36,70 @@ const join = (dispatch, logger) => {
...
+              // this handles if fork is within fork without join wrapping
+              // the inner fork
+              if (outermostTasks.info !== upStreamTasks.info) {
+                const newUpStreamTasks = taskCreationDispatcher(dispatch, upStreamTasks, task.info.uid, forkee.info.uid)
+                const newOutermostTasks = taskCreationDispatcher(dispatch, outermostTasks, task.info.uid, forkee.info.uid)
+                // notice that if not within a fork it creates a lineage
+                // with the joint tasks plus the outermostTask with a new uid
+                const lineage = [forkee].concat(newUpStreamTasks).concat(newOutermostTasks)
+                lineageGetter(dispatch, lineage, joinages)
+              } else { // if a fork is found within another fork
```
Comment (thejmazz) [inline lib/orchestrators/join.js]: is the number of nested forks supported "hardcoded" based on these nested if/else blocks?

**Comment by tiagofilipe12** [comment]
This PR now also includes new pipelines for testing multiple inputs as well as new uid generation strategy in which upstream tasks are included in uid creation. Check this [GSoC17 entry](https://github.com/bionode/GSoC17/blob/master/Journal/Week_11.md).

**Comment by tiagofilipe12** [comment]
Ok, I think it is ready to merge now, but codecov complains about code that is necessary for some of the tests I ran. Maybe you ( @thejmazz  ) can find a way to simplify the code.

**Comment by tiagofilipe12** [comment]
Also added a concurrency pipeline example using bluebird Promise.

**Comment by thejmazz** [comment]
As long as tests are passing I don't mind merging even if code coverage drops. However, if you look at the [coverage](https://codecov.io/gh/bionode/bionode-watermill/compare/7359cedaf64a2204dc7c077271fd5ac051c319cf...155635341cbcbf2889b02ee29ed82c4ef1c6b4c8/src/lib/orchestrators/join.js), it is just that `if (forkee.info.props)` and `else if (forkee.info.type === 'fork')` are not being ran in any of the tests. So if you can take a look how to resolve that, it will probably be enough to satisfy codecov.

**Review by thejmazz — APPROVED** [review]

**Pull request merged**